src/cockpit/polkit.py

```diff
@@ -100,56 +100,52 @@ async def _communicate(self, process: asyncio.subprocess.Process):
-    except KeyError:
-        logger.debug('XDG_SESSION_ID not set, not registering polkit agent')
-        return None
-
-    agent_object = org_freedesktop_PolicyKit1_AuthenticationAgent(responder)
-    agent_slot = system_bus.add_object('/', agent_object)
-
-    # register agent
-    locale_name = locale.setlocale(locale.LC_MESSAGES, '')
-    await system_bus.call_method_async(
-        'org.freedesktop.PolicyKit1',
-        '/org/freedesktop/PolicyKit1/Authority',
-        'org.freedesktop.PolicyKit1.Authority',
-        'RegisterAuthenticationAgent',
-        '(sa{sv})ss',
-        subject, locale_name, '/')
-    logger.debug('Registered agent for %r and locale %s', subject, locale_name)
-    return {'subject': subject, 'slot': agent_slot}
-
-
-async def unregister_agent(agent):
-    """Unregister polkit agent.
-
-    You have to do this explicitly (not just rely on garbage collection), otherwise
-    the next register_agent() will fail with "there is already a registered agent".
-
-    This receives the agent object returned by @register_agent().
-    """
-    system_bus = bus.Bus.default_system()
-    await system_bus.call_method_async(
-        'org.freedesktop.PolicyKit1',
-        '/org/freedesktop/PolicyKit1/Authority',
-        'org.freedesktop.PolicyKit1.Authority',
-        'UnregisterAuthenticationAgent',
-        '(sa{sv})s',
-        agent['subject'], '/')
-    agent['slot'].cancel()
-    logger.debug('Unregistered agent for %r', agent['subject'])
+    def __init__(self, responder: InteractionResponder):
+        self.responder = responder
+        self.agent_slot = None
+
+    async def __aenter__(self):
+        try:
+            self.system_bus = bus.Bus.default_system()
+        except OSError as e:
+            logger.debug('cannot connect to system bus, not registering polkit agent: %s', e)
+            return self
+
+        try:
+            # may refine that with a D-Bus call to logind
+            self.subject = ('unix-session', {'session-id': {'t': 's', 'v': os.environ['XDG_SESSION_ID']}})
+        except KeyError:
+            logger.debug('XDG_SESSION_ID not set, not registering polkit agent')
+            return self
+
+        agent_object = org_freedesktop_PolicyKit1_AuthenticationAgent(self.responder)
+        self.agent_slot = self.system_bus.add_object('/', agent_object)
+
+        # register agent
+        locale_name = locale.setlocale(locale.LC_MESSAGES, '')
+        await self.system_bus.call_method_async(
+            'org.freedesktop.PolicyKit1',
+            '/org/freedesktop/PolicyKit1/Authority',
+            'org.freedesktop.PolicyKit1.Authority',
+            'RegisterAuthenticationAgent',
+            '(sa{sv})ss',
+            self.subject, locale_name, '/')
+        logger.debug('Registered agent for %r and locale %s', self.subject, locale_name)
+        return self
+
+    async def __aexit__(self, _exc_type, _exc_value, _traceback):
+        if self.agent_slot:
```

Comment:
I think you don't need to do this conditionally anymore.  I don't think it's possible to have `__aexit__()` without `__aenter()__` and if `__aenter__()` throws then you're also not supposed to get `__aexit__()` either.

Probably makes sense to do it as an assert.

Reply:
Indeed the few tests I ran locally seem to pass, but I'm sceptical: `__aenter__` does *not* set agent_slot if there is no system D-Bus or no `XDG_SESSION_ID`, in which case agent_slot remains `None`, but there is no exception.

Reply:
I verified that this crashes, by temporarily changing do_connect_transport() to always start a polkit agent and ` env -u XDG_SESSION_ID pytest -k test_superuser`. Returning `None` in these "short-circuit" cases also fails, __aenter__ has to return something else. So this is necessary.